File: 3794-find-the-minimum-amount-of-time-to-brew-potions/3794-find-the-minimum-amount-of-time-to-brew-potions.java
class Solution {
    long minTime(int[] skill, int[] mana) {
    long[] endTime = new long[skill.length];
    Arrays.fill(endTime, 0);
    for(int i = 0; i < mana.length; ++i){
        long t = 0, maxDiff = 0;
        for(int j = 0; j < skill.length; ++j){
            maxDiff = Math.max(maxDiff, endTime[j] - t);
            t += (long)skill[j] * (long)mana[i];
            endTime[j] = t;
        }
        for(int j = 0; j < skill.length; ++j) endTime[j] += maxDiff;
    }
    return endTime[endTime.length-1];
}
}

File: 3794-find-the-minimum-amount-of-time-to-brew-potions/3794-find-the-minimum-amount-of-time-to-brew-potions.cpp
class Solution {
public:
    long long minTime(vector<int>& skill, vector<int>& mana) {
    vector<long long> endTime(skill.size(), 0);
    for(int i = 0; i < mana.size(); ++i){
        long long t = 0, maxDiff = 0;
        for(int j = 0; j < skill.size(); ++j){
            maxDiff = max(maxDiff, endTime[j] - t);
            t += (long long)skill[j] * (long long)mana[i];
            endTime[j] = t;
        }
        for(int j = 0; j < skill.size(); ++j) endTime[j] += maxDiff;
    }
    return endTime.back();
}
};